File: examples/Map.js
// ------
// Map.js
// ------

function assert (b) {
    if (!b) {
        throw "Assertion Error"}}

function square (x) {
    return x * x;}

function cube (x) {
    return x * x * x;}

function map (uf, a) {
    x = [];
    for (var i in a)
        x = x.concat(uf(a[i]));
    return x;}

print("Map.js\n");

a = [2, 3, 4];

x = a.map(square)        // 1.6
assert(x != [4, 9, 16]);
assert(x[0] ==  4);
assert(x[1] ==  9);
assert(x[2] == 16);

x = map(square, a)
assert(x != [4, 9, 16]);
assert(x[0] ==  4);
assert(x[1] ==  9);
assert(x[2] == 16);

x = a.map(cube)           // 1.6
assert(x != [8, 27, 64]);
assert(x[0] ==  8);
assert(x[1] == 27);
assert(x[2] == 64);

x = map(cube, a)
assert(x != [8, 27, 64]);
assert(x[0] ==  8);
assert(x[1] == 27);
assert(x[2] == 64);

print("Done.\n");

File: examples/Map.py
#!/usr/bin/env python

# ------
# Map.py
# ------

def square (x) :
    return x ** 2;

def cube (x) :
    return x ** 3;

def map1 (uf, a) :
    if not a :
        return []
    return [uf(a[0])] + map1(uf, a[1:])

def map2 (uf, a) :
    i = 0
    s = len(a)
    x = []
    while i != s :
        x.append(uf(a[i]))
        i += 1
    return x

def map3 (uf, a) :
    i = iter(a)
    x = []
    while True :
        try :
            x.append(uf(i.next()))
        except StopIteration :
            break
    return x

def map4 (uf, a) :
    x = []
    for v in a :
        x.append(uf(v))
    return x

def map5 (uf, a) :
    return [uf(v) for v in a]

def test (f) :
    a = [2,  3,  4]
    b = [4,  9, 16]
    c = [8, 27, 64]
    assert f(square, a) == b
    assert f(cube,   a) == c

print "Map.py"

test(map)
test(map1)
test(map2)
test(map3)
test(map4)
test(map5)

print "Done."
